frame/staking/src/lib.rs

```diff
@@ -251,46 +251,55 @@ mod mock;
 #[cfg(test)]
 mod tests;
 mod migration;
-mod slashing;
 
+pub mod slashing;
+pub mod offchain_election;
 pub mod inflation;
 
-use sp_std::{prelude::*, result};
+use sp_std::{prelude::*, result, convert::{TryInto, From}};
 use codec::{HasCompact, Encode, Decode};
 use frame_support::{
-	decl_module, decl_event, decl_storage, ensure, decl_error,
+	decl_module, decl_event, decl_storage, ensure, decl_error, debug, Parameter,
 	weights::SimpleDispatchInfo,
 	traits::{
-		Currency, LockIdentifier, LockableCurrency,
-		WithdrawReasons, OnUnbalanced, Imbalance, Get, Time
+		Currency, LockIdentifier, LockableCurrency, WithdrawReasons, OnUnbalanced, Imbalance, Get,
+		Time, EstimateNextSessionChange, OnReapAccount,
 	}
 };
-use pallet_session::historical::SessionManager;
+use pallet_session::historical;
 use sp_runtime::{
-	Perbill,
-	RuntimeDebug,
+	Perbill, RuntimeDebug, RuntimeAppPublic,
 	curve::PiecewiseLinear,
 	traits::{
 		Convert, Zero, One, StaticLookup, CheckedSub, Saturating, Bounded, SaturatedConversion,
-		SimpleArithmetic, EnsureOrigin,
-	}
+		SimpleArithmetic, EnsureOrigin, Member, OpaqueKeys
+	},
+	transaction_validity::{
+		TransactionValidityError, TransactionValidity, ValidTransaction, InvalidTransaction, TransactionPriority,
+	},
 };
 use sp_staking::{
 	SessionIndex,
 	offence::{OnOffenceHandler, OffenceDetails, Offence, ReportOffence},
 };
 #[cfg(feature = "std")]
 use sp_runtime::{Serialize, Deserialize};
-use frame_system::{self as system, ensure_signed, ensure_root};
+use frame_system::{
+	self as system, ensure_signed, ensure_root, ensure_none,
+	offchain::SubmitUnsignedTransaction,
+};
 
-use sp_phragmen::ExtendedBalance;
-use frame_support::traits::OnReapAccount;
+use sp_phragmen::{ExtendedBalance, StakedAssignment, generate_compact_solution_type};
 
 const DEFAULT_MINIMUM_VALIDATOR_COUNT: u32 = 4;
+// ------------- IMPORTANT NOTE: must be the same as `generate_compact_solution_type`.
 const MAX_NOMINATIONS: usize = 16;
 const MAX_UNLOCKING_CHUNKS: usize = 32;
 const STAKING_ID: LockIdentifier = *b"staking ";
 
+// ------------- IMPORTANT NOTE: must be the same as `MAX_NOMINATIONS`.
+generate_compact_solution_type!(pub CompactAssignments, 16);
```

Comment:
You should let `CompactAssignments` implement a trait.

This trait a const type `MAX_NOMINATIONS`. By doing it this way, we don't need to make sure that `MAX_NOMINATIONS` is updated manually.

Reply:
again, https://www.reactiongifs.us/wp-content/uploads/2013/03/mind_blown_nph.gif

Reply:
done.